app/gui2/src/util/ast/index.ts

```diff
@@ -53,47 +53,53 @@ export function readTokenSpan(token: Token, code: string): string {
 /**
  * Read direct AST children.
  */
-// TODO[ao] This is a very hacky way of implementing this. The better solution would be to generate
-// such a function from parser types schema. See parser-codegen package.
-export function* childrenAstNodes(obj: unknown): Generator<Tree> {
-  function* astNodeOrChildren(obj: unknown) {
-    if (obj instanceof Tree.AbstractBase) {
-      yield obj as Tree
+export function childrenAstNodes(obj: LazyObject): Tree[] {
+  const children: Tree[] = []
+  const visitor = (obj: LazyObject) => {
+    if (isTree(obj)) {
+      children.push(obj)
     } else {
-      yield* childrenAstNodes(obj)
-    }
-  }
-  if (obj == null) return
-  const maybeIterable = obj as Record<symbol, unknown>
-  const isIterator = typeof maybeIterable[Symbol.iterator] == 'function'
-  if (obj instanceof LazyObject) {
-    const prototype = Object.getPrototypeOf(obj)
-    const descriptors = Object.getOwnPropertyDescriptors(prototype)
-    const objWithProp = obj as unknown as Record<string, unknown>
-
-    for (const prop in descriptors) {
-      if (descriptors[prop]?.get == null) continue
-      const value = objWithProp[prop]
-      yield* astNodeOrChildren(value)
-    }
-  } else if (isIterator) {
-    const iterable = obj as Iterable<unknown>
-    for (const element of iterable) {
-      yield* astNodeOrChildren(element)
+      obj.visitChildren(visitor)
     }
   }
+  obj.visitChildren(visitor)
+  return children
+}
+
+function isTree(obj: LazyObject): obj is Tree {
+  // Every descendant of `Tree.AbstractBase` is a member of the `Tree` union.
+  return obj instanceof Tree.AbstractBase
+}
+
+function isToken(obj: LazyObject): obj is Token {
+  // Every descendant of `Token.AbstractBase` is a member of the `Token` union.
+  return obj instanceof Token.AbstractBase
```

Comment:
It would be nice to have these function defined in generated code (for any type with AbstractBase) - then the `AbstractBase` class could be not exported (as it was before). But it can also wait for some future PR if hard.